xc7/utils/prjxray_form_channels.py

```diff
@@ -72,7 +75,127 @@ def import_site_type(db, write_cur, site_types, site_type_name):
         )
 
 
-def import_tile_type(db, write_cur, tile_types, site_types, tile_type_name):
+def create_get_switch(conn):
+    """ Returns functions to get or create switches with various timing.
+
+    Every switch that requires different timing is given it's own switch
+    in VPR.  get_switch returns a switch a pip.  get_switch_timing returns
+    a switch given a particular timing.
+
+    """
+    write_cur = conn.cursor()
+
+    pip_cache = {}
+
+    write_cur.execute(
+        "SELECT pkey FROM switch WHERE name = ?",
+        ("__vpr_delayless_switch__", )
+    )
+    pip_cache[(False, 0.0, 0.0, 0.0)] = write_cur.fetchone()[0]
+
+    def get_switch_timing(
+            is_pass_transistor, delay, internal_capacitance, drive_resistance
+    ):
+        """ Return a switch that matches provided timing.
+
+        Arguments
+        ---------
+        is_pass_transistor : bool-like
+            If true, this switch should be represented as a "pass_gate".
+
+        delay : float or convertable to float
+            Intrinsic delay through switch (seconds)
+
+        internal_capacitance : float or convertable to float
+            Internal capacitance to switch (Farads).
+
+        drive_resistance : float or convertable to float
+            Drive resistance from switch (Ohms).
+
+        Returns
+        -------
+        switch_pkey : int
+            Switch primary key that represents provided arguments.
+
+        """
+        key = (
+            bool(is_pass_transistor), float(delay), float(drive_resistance),
+            float(internal_capacitance)
+        )
+
+        if key not in pip_cache:
+            name = 'routing'
+            switch_type = 'mux'
+            if is_pass_transistor:
+                name = 'pass_transistor'
+                switch_type = 'pass_gate'
+
+            name = '{}_R{}_C{}_Tdel{}'.format(
+                name, drive_resistance, internal_capacitance, delay
+            )
+
+            write_cur.execute(
+                """
+INSERT INTO switch(name, internal_capacitance, drive_resistance, intrinsic_delay, switch_type)
+VALUES
+    (?, ?, ?, ?, ?)""", (
+                    name, internal_capacitance, drive_resistance, delay,
+                    switch_type
+                )
+            )
+            pip_cache[key] = write_cur.lastrowid
+
+            write_cur.connection.commit()
+
+        return pip_cache[key]
+
+    def get_switch(pip, pip_timing):
+        """ Return switch_pkey for given pip timing.
+
+        Arguments
+        ---------
+        pip : tile.Pip object
+            Pip being modelled
+        pip_timing : tile.PipTiming object
+            Pip timing being modelled
+
+        Returns
+        -------
+        switch_pkey : int
+            Switch primary key that represents provided arguments.
+
+        """
+        delay = 0.0
+        drive_resistance = 0.0
+        internal_capacitance = 0.0
+
+        if pip_timing is not None:
```

Comment:
Would prefer if we didn't have to convert here. Could add conversion to the PipTiming class.
Higher level question, why are the values not in base SI?

Reply:
The reason I chose these units is that timing values are done in non-SI units (e.g. ns or ps) by default.  The choose units are internally consistent (milliOhms * microFarads = nanoseconds), and 21 ns is much easier to read than 0.000000021 seconds.  For model development it is significantly easier to work with units that are close to the decimal point.

It is worth noting that VPR cannot accept anything besides SI units, hence why the conversion is being done here.

Reply:
Makes sense. I think a comment in `PipTiming` describing the internal consistency and methods in the class to return in base SI would be cleaner.

Reply:
There is a comment to that affect here: https://github.com/SymbiFlow/prjxray/blob/master/prjxray/timing.py#L75 but I can add more

Reply:
How do you feel about using http://docs.enthought.com/scimath/units/unit_numpy.html ?

Reply:
I'm unimpressed with the documentation on scimath, I withdraw the suggestion.

Reply:
Ahh I was looking at [PipTiming](https://github.com/SymbiFlow/prjxray/blob/02026c2d8bffd5b0d050b0d0fa242590ba7e6aa8/prjxray/tile.py#L39)